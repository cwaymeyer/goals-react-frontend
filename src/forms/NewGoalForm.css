.NewGoalForm {
  background-color: rgb(244, 248, 252);
}
