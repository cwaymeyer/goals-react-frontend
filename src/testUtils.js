import UserContext from "./context/userContext";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

const progress = [
  {
    id: 10,
    goal_id: 1,
    weight: 50,
    reps: 5,
    orm: "56.3",
    date: "1645477469350",
  },
  {
    id: 11,
    goal_id: 1,
    weight: 50,
    reps: 6,
    orm: "58.1",
    date: "1645477669350",
  },
];

const starting_progress = {
  orm: "56.3",
  date: "1645477269350",
};

const latest_progress = {
  orm: "58.1",
  date: "1645477669350",
};

const goal = {
  id: "1",
  name: "Test Goal",
  username: "demoUser",
  start_weight: null,
  target_weight: 100,
  timeline: 3,
  start_date: "1645477269350",
  end_date: "1653163269350",
  progress: progress,
  starting_progress: starting_progress,
  latest_progress: latest_progress,
};

const demoUser = {
  username: "demoUser",
  password: "password",
  goals: [goal],
};

const UserProvider = ({ children, currentUser = demoUser }) => (
  <UserContext.Provider value={{ currentUser }}>
    {children}
  </UserContext.Provider>
);

const renderWithRouter = (ui, { route = "/" } = {}) => {
  window.history.pushState({}, "Test page", route);

  return render(ui, { wrapper: BrowserRouter });
};

export {
  UserProvider,
  renderWithRouter,
  goal,
  starting_progress,
  latest_progress,
};
