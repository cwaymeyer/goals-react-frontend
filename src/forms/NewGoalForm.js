import "./NewGoalForm.css";
import { useState, useContext } from "react";
import { useNavigate } from "react-router-dom";
import UserContext from "../context/userContext";
import getMonthsFromNow from "../helpers/getMonthsFromNow";
import getWeeksBetweenDates from "../helpers/getWeeksBetweenDates";
import {
  Container,
  Box,
  Grid,
  Button,
  FormControl,
  Select,
  InputLabel,
  TextField,
  InputAdornment,
  MenuItem,
  Alert,
} from "@mui/material";

function NewGoalForm({ createGoal }) {
  const navigate = useNavigate();
  const { currentUser } = useContext(UserContext);
  const initialState = {
    name: undefined,
    target_weight: undefined,
    timeline: undefined,
  };
  const [formData, setFormData] = useState(initialState);
  const [formErrors, setFormErrors] = useState([]);

  /** Generate goal data from form data
   *
   * Returns { name, username, start_weight, target_weight, timeline, start_date, end_date }
   */
  function setGoalData(data) {
    let goalData = {};

    goalData.name = data.name;
    goalData.username = currentUser.username;
    goalData.start_weight = null;
    goalData.target_weight = data.target_weight;

    let { currentDate, targetDate } = getMonthsFromNow(data.timeline);
    goalData.timeline = getWeeksBetweenDates(data.timeline);
    goalData.start_date = currentDate;
    goalData.end_date = targetDate;

    return goalData;
  }

  function handleChange(e) {
    const { name, value } = e.target;
    setFormData((formData) => ({ ...formData, [name]: value }));
  }

  async function handleSubmit(e) {
    e.preventDefault();
    let goalData = setGoalData(formData);
    let result = await createGoal(goalData);
    if (result.success) {
      navigate("/home");
    } else {
      setFormErrors(result.err[0]);
    }
  }

  return (
    <div className="NewGoalForm">
      <Container maxWidth="sm">
        <Box sx={{ textAlign: "center", mt: 5, mb: -2 }}>
          <h1>Set a new goal üèãÔ∏è‚Äç‚ôÄÔ∏è</h1>
        </Box>
        <form onSubmit={handleSubmit} data-testid="new-goal-form">
          <Grid container p={3}>
            <Grid item xs={12} my={1}>
              <TextField
                label="Name"
                name="name"
                value={formData.name}
                onChange={handleChange}
                sx={{ width: 1 / 1 }}
              />
            </Grid>
            <Grid item xs={6} my={1} pr={1}>
              <TextField
                label="Target weight"
                name="target_weight"
                value={formData.target_weight}
                onChange={handleChange}
                sx={{ width: 1 / 1 }}
                InputProps={{
                  endAdornment: (
                    <InputAdornment position="end">lb</InputAdornment>
                  ),
                }}
              />
            </Grid>
            <Grid item xs={6} my={1} pl={1}>
              <FormControl fullWidth>
                <InputLabel id="timeline-btn">Timeline</InputLabel>
                <Select
                  labelId="timeline-btn-label"
                  name="timeline"
                  label="Timeline"
                  onChange={handleChange}
                  id="timeline-btn"
                  sx={{ width: 1 / 1 }}
                  size="large"
                >
                  <MenuItem value={3}>3 months</MenuItem>
                  <MenuItem value={6}>6 months</MenuItem>
                  <MenuItem value={9}>9 months</MenuItem>
                  <MenuItem value={12}>1 year</MenuItem>
                </Select>
              </FormControl>
            </Grid>
            <Button
              type="submit"
              variant="contained"
              color="primary"
              onSubmit={handleSubmit}
              sx={{ mt: 2 }}
            >
              Create goal
            </Button>
            <Button
              variant="text"
              color="secondary"
              onClick={() => navigate("/home")}
              sx={{ mt: 2, ml: 2 }}
            >
              Back
            </Button>
          </Grid>
        </form>
        {formErrors.length > 0 ? (
          <Alert severity="error">{formErrors}</Alert>
        ) : null}
      </Container>
    </div>
  );
}

export default NewGoalForm;
