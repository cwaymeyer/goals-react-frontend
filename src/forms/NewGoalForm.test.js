import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import { UserProvider } from "../testUtils";
import NewGoalForm from "./NewGoalForm";

it("renders without crashing", function () {
  render(
    <UserProvider>
      <MemoryRouter>
        <NewGoalForm />
      </MemoryRouter>
    </UserProvider>
  );
});
