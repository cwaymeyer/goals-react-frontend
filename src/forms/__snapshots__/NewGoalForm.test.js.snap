// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="NewGoalForm"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
    >
      <div
        class="MuiBox-root css-n3fmse"
      >
        <h1>
          Set a new goal üèãÔ∏è‚Äç‚ôÄÔ∏è
        </h1>
      </div>
      <form>
        <div
          class="MuiGrid-root MuiGrid-container css-1cs2yco-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1qobofj-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="mui-3"
                id="mui-3-label"
              >
                Name
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-3"
                  name="name"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      Name
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1aga215-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="mui-4"
                id="mui-4-label"
              >
                Target weight
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-4"
                  name="target_weight"
                  type="text"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                  >
                    lb
                  </p>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      Target weight
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-ri2eq3-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                id="timeline-btn"
              >
                Timeline
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-vwjx0e-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="timeline-btn-label timeline-btn"
                  class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="timeline-btn"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="notranslate"
                  >
                    ‚Äã
                  </span>
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  name="timeline"
                  tabindex="-1"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      Timeline
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-lll1vm-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Create goal
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1tpx81f-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Back
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
